### ISSUE TITLE
Revisit count queries for LDAPStorageProvider

### ISSUE THREAD

**Issue by vramik** [keycloak/keycloak#20156]
https://github.com/keycloak/keycloak/pull/19050#discussion_r1184810342

**Comment by mhajas** [comment]
@mposolda This is a follow-up on [your comment](https://github.com/keycloak/keycloak/pull/19050#pullrequestreview-1430181658):
> As I guess that for example using UsersResource.getUsersCount (but likely also other methods) can have significant performance impact with the new behaviour?

I would like to understand the reasons for the significant performance impact of the new behaviour. The previous version also read users in count methods including importing them to local storage. We changed it to read page by page so it can result to loading the whole LDAP, is this the only reason for performance impact? Or is there something more from your point of view?

Also, do you see any problem setting the count methods return value to a fixed value of 100 in the LDAP provider? The reason for the fixed value is that it will be much faster and we would not read from LDAP twice from this code (which seems to be the only place that is calling count query from LDAP):
https://github.com/keycloak/keycloak/blob/33215ab6f498d20bc8420d220e1a6b0de22188fd/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L202-L260

**Comment by mposolda** [comment]
> @mposolda This is a follow-up on [your comment](https://github.com/keycloak/keycloak/pull/19050#pullrequestreview-1430181658):
> 
> > As I guess that for example using UsersResource.getUsersCount (but likely also other methods) can have significant performance impact with the new behaviour?
> 
> I would like to understand the reasons for the significant performance impact of the new behaviour. The previous version also read users in count methods including importing them to local storage. We changed it to read page by page so it can result to loading the whole LDAP, is this the only reason for performance impact? Or is there something more from your point of view?

Hmm... AFAIK in current Keycloak main, when you call something like:
```
adminClient.realm("foo").users().count();
```
it will return only users, who are already imported in localStorage. It won't do any imports from LDAP storage. It will just call `LDAPStorageProvider.getUsersCount(realm)`, which in current main is hardcoded to return 0.

With https://github.com/keycloak/keycloak/pull/19050/, it seems that `LDAPStorageProvider.getUsersCount(realm)` would call `searchLDAPByAttributes(realm, params, null, null).filter(filterLocalUsers(realm)).count()`, which looks like pretty much searching whole LDAP tree? I could be wrong in something from the above (you can doublecheck...).

But I don't see this as a big issue as AFAIK new admin console doesn't call `count()` endpoint. But still, it worth to be documented if we keep it this way IMO?

> 
> Also, do you see any problem setting the count methods return value to a fixed value of 100 in the LDAP provider? The reason for the fixed value is that it will be much faster and we would not read from LDAP twice from this code (which seems to be the only place that is calling count query from LDAP):
> 
> https://github.com/keycloak/keycloak/blob/33215ab6f498d20bc8420d220e1a6b0de22188fd/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L202-L260

I don't see this as an issue, but we should still likely document it in the upgrading guide? As some people can still directly call the `count()` endpoint from admin REST API. And for those users, we should likely document that it's either behaviour change (due the hardcoded value of 100) or performance impact (if we kept like in Vlasta's PR). 

I don't see big issue in any of this as long as it is documented if it makes life easier for you guys with new storage :-)

**Comment by vramik** [comment]
Thanks @mposolda!

regarding a calling `adminClient.realm("foo").users().count();`. AFAIK this call first lands in `UserStorageManager` and there is queried only local storage: https://github.com/keycloak/keycloak/blob/893a285ac483a768bd304c82e6af56f7b0e06785/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L397-L420

If I didn't miss something it should be fine in this regard.

> With https://github.com/keycloak/keycloak/pull/19050, it seems that LDAPStorageProvider.getUsersCount(realm) would call searchLDAPByAttributes(realm, params, null, null).filter(filterLocalUsers(realm)).count(), which looks like pretty much searching whole LDAP tree? I could be wrong in something from the above (you can doublecheck...).

We decided to split the work and address exactly this issue in separate PR, which should also target 22.0.0 milestone. There is draft PR sent with Michal's suggestion: https://github.com/keycloak/keycloak/pull/20447/commits/fe6dd5616c6f94910cf23ea2e7bef1a464e2debf

I'll add a note to the upgrading guide within https://github.com/keycloak/keycloak/pull/20447 regarding hardcoded value.

**Comment by mhajas** [comment]
> it will return only users, who are already imported in localStorage. It won't do any imports from LDAP storage. It will just call `LDAPStorageProvider.getUsersCount(realm)`, which in current main is hardcoded to return 0.

It will return only from local storage because `UserStorageManager` does not add LDAP to the result:
https://github.com/keycloak/keycloak/blob/893a285ac483a768bd304c82e6af56f7b0e06785/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L392-L420

The count methods are defaulting to querying and LDAP does not override them all:
https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java#L105

see
https://github.com/keycloak/keycloak/blob/893a285ac483a768bd304c82e6af56f7b0e06785/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java#L344-L347

> With https://github.com/keycloak/keycloak/pull/19050, it seems that LDAPStorageProvider.getUsersCount(realm) would call searchLDAPByAttributes(realm, params, null, null).filter(filterLocalUsers(realm)).count(), which looks like pretty much searching whole LDAP tree? I could be wrong in something from the above (you can doublecheck...).

Yes, it is, but the way I understand it is that prior to #19050 we loaded some default maximum of users (depending on LDAP implementation?) and we imported all of them into the local database. Please correct me if I am wrong.

However, after #19050 we changed the search to dynamically load pages based on the needed therefore the LDAP maximum is not applied and therefore we load all users from LDAP when executing count. 

And this ^ is what we need to fix as part of this issue.

**Comment by mposolda** [comment]
@mhajas @vramik Hmm.... I've tried to add this method in `LDAPProvidersIntegrationTest` (using current Keycloak main):
```
    @Test
    public void testSyncRegistrationForceDefaultt() {
        int count = testRealm().users().count();
        System.out.println("Count: " + count);
    }
```

Running this would propagate to `LDAPStorageProvider.getUsersCount(RealmModel realm)` . It actually uses this method in UserStorageManager, which delegates to other providers : https://github.com/keycloak/keycloak/blob/893a285ac483a768bd304c82e6af56f7b0e06785/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L383-L389

The stacktrace looks like this:
```
	at org.keycloak.storage.ldap.LDAPStorageProvider.getUsersCount(LDAPStorageProvider.java:346)
	at org.keycloak.storage.UserStorageManager.lambda$getUsersCount$21(UserStorageManager.java:386)
	at org.keycloak.utils.ServicesUtils.lambda$timeBoundOne$1(ServicesUtils.java:83)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
	at org.keycloak.storage.UserStorageManager.getUsersCount(UserStorageManager.java:387)
	at org.keycloak.models.cache.infinispan.UserCacheSession.getUsersCount(UserCacheSession.java:555)
	at org.keycloak.storage.user.UserQueryProvider.getUsersCount(UserQueryProvider.java:52)
	at org.keycloak.services.resources.admin.UsersResource.getUsersCount(UsersResource.java:413)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

In current Keycloak main, `LDAPStorageProvider.getUSersCount(realm)` is hardcoded to 0. So there seems not any search call to LDAP server (and hence not any import from LDAP server) when calling this `count()` admin REST API without arguments.

But with Vlasta's commit with hardcoded values https://github.com/keycloak/keycloak/pull/20447/commits/759098007ce200b7257751ebc478dbdb2b81714b , which also has this documented, we should be fine IMO. Feel free to use this commit as a follow-up to https://github.com/keycloak/keycloak/pull/19050 or directly add it to https://github.com/keycloak/keycloak/pull/19050 PR. Whatever works better. As long as we have this change in Keycloak 22, we should be fine IMO.

**Comment by mhajas** [comment]
Thank you @mposolda, I completely missed that one method looking into providers in `UserStorageManager`. 

In that case, we indeed need to return 0 from that method in `LDAPStorageProvider`.
